### PR TITLE
[DataGridPremium] Server-side data source with row grouping

### LINKED ISSUE #10859: [data grid] Implement server-side data source in row grouping
This issue covers the server-side row-grouping (based on the data source).
Feel free to upvote the issue to see it land faster.

### PR DESCRIPTION
**Pull request by MBilalShafi**
👨‍💻 Ready for Review 👨‍💻

Part of #8179
Resolves #10859

**Preview:** https://deploy-preview-13826--material-ui-x.netlify.app/x/react-data-grid/server-side-data/row-grouping/

### Action items

- [x] Add support for `groupingValueGetter`
- [x] Support row grouping features on `useMockServer`:
  - [x] Sorting 
  - [x] Filtering
  - [x] Pagination
  - [x] Support more datasets
- [x] Rows with missing groups
- [x] Fix Argos' regression
- [x] Add/refine documentation

### Follow-ups:

- [ ] Support sort and filter on the Group column(s) for row grouping and tree data
- [ ] Test coverage

### CODE REVIEW

**Comment by tommy-wl** [comment]
Hi @MBilalShafi, are you able to provide an ETA for this? Thank you!

**Comment by MBilalShafi** [comment]
Hey @tommy-wl, thank you for reaching out.

The ETA for this feature is around the end of Q3, its currently one of the main focus areas for the team. Feel free to test the under development feature and provide some early feedback if you want.

**Comment by forwardslahsdotj** [comment]
@MBilalShafi Looking for an update on this or an ETA. Premium customer - Support ID: 96081

**Comment by MBilalShafi** [comment]
@forwardslahsdotj

The goal is to deliver the feature by around Q3 end, though it depends on how fast the closure of remaining action items and the code review process takes. I hope it gets closed before the v8 pre-release phase, which starts sometime in October.

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx [hunk 1]
```diff
@@ -122,6 +135,7 @@ interface CreateGroupingColDefMonoCriteriaParams {
    * This value comes `prop.groupingColDef`.
    */
   colDefOverride: GridGroupingColDefOverride | null | undefined;
+  strategy?: typeof ROW_GROUPING_STRATEGY | typeof DATA_SOURCE_ROW_GROUPING_STRATEGY;
```
Comment (romgrk) [inline packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx:138]: Why not something like `mode: 'client' | 'server'` instead?

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx [hunk 2]
```diff
@@ -12,10 +12,13 @@ import { GridColumnRawLookup, isSingleSelectColDef } from '@mui/x-data-grid-pro/
 import { GridApiPremium } from '../../../models/gridApiPremium';
 import { GridGroupingColumnFooterCell } from '../../../components/GridGroupingColumnFooterCell';
 import { GridGroupingCriteriaCell } from '../../../components/GridGroupingCriteriaCell';
+import { GridDataSourceGroupingCriteriaCell } from '../../../components/GridDataSourceGroupingCriteriaCell';
 import { GridGroupingColumnLeafCell } from '../../../components/GridGroupingColumnLeafCell';
 import {
   getRowGroupingFieldFromGroupingCriteria,
   GRID_ROW_GROUPING_SINGLE_GROUPING_FIELD,
+  ROW_GROUPING_STRATEGY,
+  DATA_SOURCE_ROW_GROUPING_STRATEGY,
```
Comment (romgrk) [inline packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx:21]: `ROW_GROUPING_STRATEGY_DATA_SOURCE` would be easier to read.

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx [hunk 3]
```diff
@@ -73,9 +74,6 @@ export const useGridRowGrouping = (
     changeEvent: 'rowGroupingModelChange',
   });
 
-  /**
-   * API METHODS
-   */
```
Comment (romgrk) [inline packages/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx:78]: I kinda like having these, I'd just replace the `**` if this change is for the jsdoc lint.

File: packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts [hunk 4]
```diff
@@ -14,15 +15,32 @@ const computeStartEnd = (paginationModel: GridPaginationModel) => {
...
+const EMPTY_ARRAY: string[] = [];
+
+const gridRowGroupingModelSelector = (state: GridStateProWithRowGrouping) =>
+  state.rowGrouping?.model ?? EMPTY_ARRAY;
+
+export const gridRowGroupingSanitizedModelSelector = createSelector(
+  gridRowGroupingModelSelector,
+  gridColumnLookupSelector,
+  (model, columnsLookup) => model.filter((field) => !!columnsLookup[field]),
+);
```
Comment (romgrk) [inline packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts:33]: For my understanding, in which cases is the sanitization necessary?

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx (hunk 3 above)
Comment (MBilalShafi) [reply packages/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx:78]: Hmm, do you think we should settle on a specific format around this or just a one-liner comment instead of JSDoc style would be good to go? Something like:
```tsx
  // API METHODS
```

File: packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts (hunk 4 above)
Comment (romgrk) [reply packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts:33]: And shouldn't this be a memoized selector, considering that it's building a derived array value?

Comment (MBilalShafi) [reply packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts:33]: When a field is provided in `rowGroupingModel` that doesn't exist (or got removed from) the column definitions.
It's similar to the one we are using in the actual `rowGrouping` [implementation](https://github.com/mui/mui-x/blob/153bce3cd50df281801ab8a4721d3aa43a919023/packages/x-data-grid-premium/src/hooks/features/rowGrouping/gridRowGroupingSelector.ts#L12-L16).

Side topic: I initially thought to move this one in the Pro package and export from `/internals` to reuse but went ahead and duplicated it. Would reuse approach will be better/less confusing?

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx (hunk 3 above)
Comment (romgrk) [reply packages/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx:78]: I'd keep it as it is, the other files already follow the same structure with the multiline comment, just remove the jsdoc marker.

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx (hunk 2 above)
Comment (MBilalShafi) [reply packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx:21]: I tried to align the [one](https://github.com/mui/mui-x/blob/153bce3cd50df281801ab8a4721d3aa43a919023/packages/x-data-grid-pro/src/hooks/features/serverSideTreeData/useGridDataSourceTreeDataPreProcessors.tsx#L38) with the tree data. Do you think we should rename both?
Tbh, the current one looks more readable to me personally, but I'm fine to rename if it looks less readable to other too. 👍

File: packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts (hunk 4 above)
Comment (romgrk) [reply packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts:33]: If it makes sense, yes I would avoid duplication.

Comment (MBilalShafi) [reply packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts:33]: You are right, it should be memoized 👍 
I'll do either of:
1. Move the original selectors to Pro and reuse them in both places.
2. Convert these to memoized 

Please vote if you have a preference:
👍 = 1, 🚀 = 2

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx (hunk 2 above)
Comment (romgrk) [reply packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx:21]: 1.
```
  ROW_GROUPING_STRATEGY,
  DATA_SOURCE_ROW_GROUPING_STRATEGY,
```

2.
```
  ROW_GROUPING_STRATEGY,
  ROW_GROUPING_STRATEGY_DATA_SOURCE,
```

I find option 2 more readable when constants are used like enum values, imho it makes it apparent that both values have the same "type".

Comment (romgrk) [reply packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx:21]: And tbh I would actually rename to something like this:

```
  ROW_GROUPING_STRATEGY_DEFAULT,
  ROW_GROUPING_STRATEGY_DATA_SOURCE,
```

File: packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts (hunk 4 above)
Comment (romgrk) [reply packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts:33]: No preference, do what you think is best.

Comment (MBilalShafi) [reply packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts]: Ended up converting the selector to memoized as bringing every related thing to the lower package (like `GridStatePremium` interface) seems a bit too much.

File: docs/data/data-grid/server-side-data/row-grouping.md [hunk 5]
```diff
@@ -2,14 +2,90 @@
...
+  },
+  getChildrenCount: (row) => {
+    // Return the number of children for the row
+    return row.childrenCount;
+  },
+};
+```
 
-👍 Upvote [issue #10859](https://github.com/mui/mui-x/issues/10859) if you want to see it land faster.
+In addition to `groupKeys`, the `getRows` callback receives a `groupFields` parameter. This corresponds to the current `rowGroupingModel`. Use `groupFields` on the server to group the data for each `getRows` call.
```
[the remark this answers is not in the document]
Comment (MBilalShafi) [reply docs/data/data-grid/server-side-data/row-grouping.md]: You mean `tree data`, right?
Yes definitely makes sense. Added!

File: packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts [hunk 6]
```diff
@@ -14,15 +19,34 @@ const computeStartEnd = (paginationModel: GridPaginationModel) => {
...
   gridSortModelSelector,
   gridPaginationModelSelector,
-  (filterModel, sortModel, paginationModel) => {
+  gridRowGroupingSanitizedModelSelector,
+  (filterModel, sortModel, paginationModel, rowGroupingModel) => {
     return {
       groupKeys: [],
-      // TODO: Implement with `rowGrouping`
-      groupFields: [],
+      groupFields: rowGroupingModel,
```
[the remark this answers is not in the document]
Comment (MBilalShafi) [reply packages/x-data-grid-pro/src/hooks/features/dataSource/gridDataSourceSelector.ts]: Ended up [using](https://github.com/mui/mui-x/pull/13826/commits/8c623d04ee4a7fd1016cd1422316faaa9b3f4010) pipe processing for it, it seems a better choice.

File: docs/data/data-grid/server-side-data/tree-data.md [hunk 7]
```diff
@@ -27,6 +33,25 @@ const customDataSource: GridDataSource = {
 };
 ```
 
+Apart from the other parameters like `filterModel`, `sortModel`, and `paginationModel`, the `getRows()` callback receives a `groupKeys` parameter. This corresponds to the keys provided for each nested level in `getGroupKey()`. Use `groupKeys` on the server to fetch the rows for each nested level.
```
Comment (MBilalShafi) [inline docs/data/data-grid/server-side-data/tree-data.md:36]:
```suggestion
Like the other parameters such as `filterModel`, `sortModel`, and `paginationModel`, the `getRows()` callback receives a `groupKeys` parameter that corresponds to the keys provided for each nested level in `getGroupKey()`.
Use `groupKeys` on the server to extract the rows for a given nested level.
```

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/useGridDataSourceRowGroupingPreProcessors.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      props.unstable_dataSource,
+      props.defaultGroupingExpansionDepth,
+      props.isGroupExpandedByDefault,
+    ],
+  );
+
+  const filterRows = React.useCallback<GridStrategyProcessor<'filtering'>>(() => {
+    const rowTree = gridRowTreeSelector(apiRef);
+
+    return skipFiltering(rowTree);
```
Comment (arminmeh) [inline packages/x-data-grid-premium/src/hooks/features/rowGrouping/useGridDataSourceRowGroupingPreProcessors.ts:111]: what is preventing you reseting the grid and making a new request with the new parameters?

Comment (MBilalShafi) [reply packages/x-data-grid-premium/src/hooks/features/rowGrouping/useGridDataSourceRowGroupingPreProcessors.ts:111]: Not sure if I follow 🤔. This is the client-side strategy processor for filtering. Since we don't have client side filtering with the `dataSource`, we skip it.

Just like for tree data:
https://github.com/mui/mui-x/blob/46df9666a5d629548f68f5fe3f6c10b7a31248b1/packages/x-data-grid-pro/src/hooks/features/serverSideTreeData/useGridDataSourceTreeDataPreProcessors.tsx#L200-L210

Did I get your point right?

Comment (arminmeh) [reply packages/x-data-grid-premium/src/hooks/features/rowGrouping/useGridDataSourceRowGroupingPreProcessors.ts:111]: ah yes, ignore this
wrong line
I was looking at the point about the group column sorting and actually wanted to comment this on the group cell definition, but ended up here 😅
I will resolve this and make the same comment there

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx [hunk 9]
```diff
@@ -25,11 +28,24 @@ const GROUPING_COL_DEF_DEFAULT_PROPERTIES: Omit<GridColDef, 'field'> = {
...
   editable: false,
   groupable: false,
 };
 
+const GROUPING_COL_DEF_FORCED_PROPERTIES_DATA_SOURCE: Pick<
+  GridColDef,
+  'type' | 'editable' | 'groupable' | 'filterable' | 'sortable' | 'aggregable'
+> = {
+  ...GROUPING_COL_DEF_FORCED_PROPERTIES_DEFAULT,
+  // TODO: Support these features on the grouping column(s)
```
Comment (arminmeh) [inline packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx:44]: what is preventing you reseting the grid and making a new request with the new parameters?

Comment (MBilalShafi) [reply packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx:44]: Thanks for the question, I wanted to drop a comment on this but forgot.

We need to comeup with some standard way to communicate with the backend server that a specific operation (sorting/filtering/etc) is requested on a certain grouping column. Currently, hardcoded field names are used for these columns (both in tree data and row grouping) that are more frontend specific.

Tree data grouping field: `__tree_data_group__`
Row grouping single grouping column: `__row_group_by_columns_group__`
Row grouping multiple grouping columns: `__row_group_by_columns_group_${field}__`

Currently when sorting or filtering, it would yield into:

```tsx
const filterModel = {
  items: [
    {
      field: '__row_group_by_columns_group_company__',
      value: 'John',
      value: 'contains',
    },
    {
      field: '__tree_data_group__',
      value: 'John',
      value: 'startsWith',
    },
  ],
};

const sortModel = [
  {
    field: '__row_group_by_columns_group__',
    sort: 'asc',
  },
  {
    field: '__tree_data_group__',
    sort: 'asc',
  },
];
```

We need to translate these values in to some generic readable values based on the current configuration before passing them on to the data source. For example when there is a specific filtering criteria defined for single column multiple row grouping, we need to communicate it to the backend, a possible field name to use for it could be `group_${criteriaField}`

That's the reason I wanted to handle it separately, to gather some community/internal feedback on the naming first and also because it'd not only cover the row grouping but also tree data.

I'll create a separate issue for this and handle it afterwards.

File: packages/x-data-grid-premium/src/components/GridDataSourceGroupingCriteriaCell.tsx [hunk 10]
```diff
@@ -0,0 +1,148 @@
+import * as React from 'react';
+import { unstable_composeClasses as composeClasses } from '@mui/utils';
+import Box from '@mui/material/Box';
+import CircularProgress from '@mui/material/CircularProgress';
+import Badge from '@mui/material/Badge';
```
Comment (romgrk) [inline packages/x-data-grid-premium/src/components/GridDataSourceGroupingCriteriaCell.tsx:5]: Replace with `rootProps.slots.baseBadge` if #15013 gets merged first.

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx [hunk 11]
```diff
@@ -122,6 +138,7 @@ interface CreateGroupingColDefMonoCriteriaParams {
    * This value comes `prop.groupingColDef`.
    */
   colDefOverride: GridGroupingColDefOverride | null | undefined;
+  strategy?: typeof ROW_GROUPING_STRATEGY_DEFAULT | typeof ROW_GROUPING_STRATEGY_DATA_SOURCE;
```
Comment (romgrk) [inline packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx:141]: Using a TS enum would be a better DX:

```typescript
enum RowGroupingStrategy {
  Default = 'default',
  DataSource = 'data-source',
}
```

File: packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx (hunk 9 above)
Comment (arminmeh) [reply packages/x-data-grid-premium/src/hooks/features/rowGrouping/createGroupingColDef.tsx:43]: thanks for all the details
yes, that is fine
let's do it afterwards

**Review by arminmeh — APPROVED** [review]

File: docs/data/data-grid/server-side-data/row-grouping.md [hunk 12]
```diff
@@ -2,14 +2,90 @@
...
+The group expansion works similar to the [data source tree data](/x/react-data-grid/server-side-data/tree-data/#group-expansion).
+The following demo uses `defaultGroupingExpansionDepth='-1'` to expand all the groups.
+
+{{"demo": "ServerSideRowGroupingGroupExpansion.js", "bg": "inline"}}
+
+## Demo
+
+In the following demo, use the auto generated data based on the `Commodities` dataset to simulate the server-side row grouping.
+
+{{"demo": "ServerSideRowGroupingFullDataGrid.js", "bg": "inline"}}
```
Comment (cherniavskii) [inline docs/data/data-grid/server-side-data/row-grouping.md:85]: Any ideas about what is causing these lags? Maybe client-side filtering kicks in?


https://github.com/user-attachments/assets/48f6190a-096e-479b-83eb-dd7ffc3e93fb

Comment (MBilalShafi) [reply docs/data/data-grid/server-side-data/row-grouping.md:85]: Thank you @cherniavskii for pointing it out. It really does look bad.

Filtering logic on the `useMockServer` side isn't very optimal atm. Results on 20k rows:

image

We'll have to replicate/reuse the frontend filtering engine there somehow to be as performant. (I guess the same mock server computation bottle-neck was a rationale for [limiting](https://github.com/mui/mui-x/blob/61bef1e4cfdc8c026eb73b7ce2887a1d459d444b/packages/x-data-grid-generator/src/services/tree-data-generator.ts#L44) tree data max row size to `1000`)

For now, I [limited](https://github.com/mui/mui-x/pull/13826/commits/d849c385fc405d485175c9104eb10ebe6638670e) the data size for this demo, it's smoother. If needed, we could try to optimize the backend logic sometime in the future.

Comment (cherniavskii) [reply docs/data/data-grid/server-side-data/row-grouping.md:85]: Oh, we didn't anticipate that the mock server wouldn't be performant enough for large datasets.
At this point, I feel like it's better to set up a backend with a database for it.

File: docs/data/data-grid/server-side-data/ServerSideRowGroupingDataGrid.js [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+      getGroupKey: (row) => row.group,
+      getChildrenCount: (row) => row.descendantCount,
+    };
+  }, [fetchRows]);
+
+  const initialState = useKeepGroupedColumnsHidden({
+    apiRef,
+    initialState: {
+      rowGrouping: {
+        model: ['company', 'director'],
```
Comment (cherniavskii) [inline docs/data/data-grid/server-side-data/ServerSideRowGroupingDataGrid.js:44]: Grouping by `year` doesn't seem to work in this demo.
Is this also related to the mock server?

File: docs/data/data-grid/server-side-data/row-grouping.md (hunk 12 above)
Comment (MBilalShafi) [reply docs/data/data-grid/server-side-data/row-grouping.md:85]: > I feel like it's better to set up a backend with a database for it.

Is it worth the effort though? Technically we could replicate the same filtering engine on the backend which we have on the client app.

Even demonstrating on a lower data size should be fine for now, since the users should already have their backend servers setup.

Is there a specific aspect in your mind we could not cover by a dummy server?

Comment (cherniavskii) [reply docs/data/data-grid/server-side-data/row-grouping.md:85]: > Is it worth the effort though? Technically we could replicate the same filtering engine on the backend which we have on the client app.

I think the effort would be smaller than making `useMockServer` work and perform well. Plus, users could use this example as a reference when integrating with their DB, as opposed to `useMockServer` which is an internal implementation that is not useful to anyone integrating Data Grid with their backend.

> Is there a specific aspect in your mind we could not cover by a dummy server?

My main concern is that we're trying to build data transformation from scratch, while we can use a database query that is optimized for it already – e.g. for row grouping, we can use groupBy + where instead of a ton of custom code: https://github.com/mui/mui-x/pull/13826/files#diff-97d37099273d5d313dc00f50320ced3b902a089d924b2092213d2573ee5c63c8

File: packages/x-data-grid/src/components/cell/GridBooleanCell.tsx [hunk 14]
```diff
@@ -54,6 +54,10 @@ function GridBooleanCellRaw(props: GridBooleanCellProps) {
     [rootProps.slots.booleanCellFalseIcon, rootProps.slots.booleanCellTrueIcon, value],
   );
 
+  if (value === undefined) {
+    return null;
+  }
```
Comment (cherniavskii) [inline packages/x-data-grid/src/components/cell/GridBooleanCell.tsx:59]: Not sure we should do this now, it's a breaking change (currently, `undefined` value is treated as `false`)

File: docs/data/data-grid/server-side-data/ServerSideRowGroupingDataGrid.js (hunk 13 above)
Comment (MBilalShafi) [reply docs/data/data-grid/server-side-data/ServerSideRowGroupingDataGrid.js:44]: Yes, [fixed](https://github.com/mui/mui-x/pull/13826/commits/1b5bd369df398b418867f0a4bf47e73c5e402bcf).

**Review by cherniavskii — APPROVED** [review]

File: docs/data/data-grid/server-side-data/row-grouping.md (hunk 12 above)
Comment (MBilalShafi) [reply docs/data/data-grid/server-side-data/row-grouping.md:85]: Your points make sense 👍

I am more concerned about the BE server infra management (server, db, etc) in this case that'd support high bandwidth without causing disruptions. I'm not sure if it will be possible using toolpad, even if it is, we have used it for internal tools for now, not on the public documentation, or have we? 🤔

In any case, let's take it as a separate issue: https://github.com/mui/mui-x/issues/15098

Comment (cherniavskii) [reply docs/data/data-grid/server-side-data/row-grouping.md:85]: > I am more concerned about the BE server infra management (server, db, etc) in this case that'd support high bandwidth without causing disruptions.

Yes, this might be challenging indeed.

File: docs/data/data-grid/row-grouping/row-grouping.md [hunk 15]
```diff
@@ -252,6 +256,10 @@ Use the `setRowChildrenExpansion` method on `apiRef` to programmatically set the
 
 {{"demo": "RowGroupingSetChildrenExpansion.js", "bg": "inline", "defaultCodeOpen": false}}
 
+:::warning
+The `apiRef.current.setRowChildrenExpansion` method is not compatible with the [server-side tree data](/x/react-data-grid/server-side-data/tree-data/) and [server-side row grouping](/x/react-data-grid/server-side-data/row-grouping/). Use `apiRef.current.unstable_dataSource.fetchRows` instead.
```
Comment (joserodolfofreitas) [inline docs/data/data-grid/row-grouping/row-grouping.md:260]: Curious, why it is not compatible? Could it call the data source to apply the behavior or is there a strong feasibility blocker?
(Question applies to both get and set methods)

**Review by joserodolfofreitas — APPROVED** [review]
Solid feature! Great work!

File: packages/x-data-grid/src/components/cell/GridBooleanCell.tsx (hunk 14 above)
Comment (MBilalShafi) [reply packages/x-data-grid/src/components/cell/GridBooleanCell.tsx]: I added this to avoid this:
image

For client-side row grouping we have a concept of autogenerated group nodes and we handle it like this:
https://github.com/mui/mui-x/blob/ab4823da20defea849b98e006dcd35ffec8d65fb/packages/x-data-grid/src/components/cell/GridBooleanCell.tsx#L145-L147

In server-side row grouping, all the rows are regular group rows and nothing is auto generated, so the above check doesn't work. Did a hacky'ish [solution](https://github.com/mui/mui-x/pull/13826/commits/10747b1958906d5d7e88ea72de046bf8656b0ca0), let me know how does it feel.

File: docs/data/data-grid/row-grouping/row-grouping.md (hunk 15 above)
Comment (MBilalShafi) [reply docs/data/data-grid/row-grouping/row-grouping.md:260]: Data source changes a bit how expansion works.
When the user expands a row, `fetchRows` is called, which fetches the rows internally and only on success it calls `setRowChildrenExpansion` internally which only updates the `isExpanded` state of the respective `rowNode`.

If the user tries to call `setRowChildrenExpansion` there will be no rows to show, only the expand icon will shift to expanded form, thus `setRowChildrenExpansion` is no longer recommended to use with the data source.

Comment (cherniavskii) [reply docs/data/data-grid/row-grouping/row-grouping.md:260]: Good point @joserodolfofreitas 
@MBilalShafi Couldn't we potentially update `setRowChildrenExpansion` to fetch rows if the data source is being used?

File: docs/data/data-grid/row-grouping/row-grouping.md
Comment (cherniavskii) [inline docs/data/data-grid/row-grouping/row-grouping.md:1]: @MBilalShafi Any ideas on how to make Codesandbox CI pass? https://ci.codesandbox.io/status/mui/mui-x/pr/13826/builds/554134
After merging this PR it will be failing for all PRs, e.g. https://github.com/mui/mui-x/pull/15108

**Comment by tommy-wl** [comment]
@MBilalShafi great work! Super excited to start using this feature as soon as it is released!

Quick question: does this work for multi-level row grouping as well?

Thanks

File: docs/data/data-grid/row-grouping/row-grouping.md
Comment (MBilalShafi) [reply docs/data/data-grid/row-grouping/row-grouping.md:1]: We need to do https://github.com/mui/mui-x/pull/15109/commits/75f3818d0036989dea8f2407c9de4a9119a30fe9 in master.
I'll open a PR soon.

I'm still figuring out why this causes issue in build only rather than `pnpm typescript`.

**Comment by MBilalShafi** [comment]
Hey @tommy-wl,

> just to confirm, does this work grouping multiple columns at the same time as well? [mui.com/x/react-data-grid/row-grouping#multiple-grouping-columns](https://mui.com/x/react-data-grid/row-grouping/#multiple-grouping-columns)

Yes, single column and multiple column grouping both should work with this feature.

File: docs/data/data-grid/row-grouping/row-grouping.md (hunk 15 above)
Comment (joserodolfofreitas) [reply docs/data/data-grid/row-grouping/row-grouping.md:260]: I missed the conversation here. 
Should we create an issue to make these approaches compatible?

Comment (MBilalShafi) [reply docs/data/data-grid/row-grouping/row-grouping.md:260]: > Couldn't we potentially update setRowChildrenExpansion to fetch rows if the data source is being used?

We certainly could, but it hasn't been done in the current implementation because `setRowChildrenExpansion` sounds like (or anticipated as) a state setter (or a pure function), hiding the `fetch` part behind would be less aligned with a user's expectation as it could add fetch delays and request failures changing it's current behavior (impure).

Using it from within `fetchRows` makes sense if we understand it this way, or we may have to change its current perception.

What do you think?

**Comment by tommy-wl** [comment]
Hi @MBilalShafi 

Server side row grouping was fine on MUI Data Grid v7 (Grouping on single and multiple columns).

Upon upgrading to v8 (8.5.2 -- also tried 8.0.0), the following was observed:

Screenshot 2025-06-16 230250

`Multiple` is the first level grouping.

When clicking on it to expand, the second level groups do appear (`Ahmed`, `Client Anastasia` and `Guest`).

However:
- Second level groups are not indented **(Issue 1)**
- First level group clicked is not expanded (chevron is still pointing to the right) **(Issue 2)**

Clicking the first level group `Multiple` again correctly shows the expanded state (chevron down) and indent for second level groups.

When trying to click and expand other first level groups, the following error appears **(Issue 3)**: 
```
Uncaught TypeError: Cannot read properties of undefined (reading 'depth')
    at gridFilterSelector.js:92:39
    at Array.reduce (<anonymous>)
    at gridFilterSelector.js:90:30
    at recomputationWrapper (createSelectorCreator.ts:392:63)
    at memoized (lruMemoize.ts:217:20)
    at dependenciesChecker (createSelectorCreator.ts:412:39)
    at memoized (weakMapMemoize.js:100:21)
    at Object.selector (createSelector.js:82:14)
    at useGridSelector.js:59:37
    at Store.js:19:35
```

These issues never occurred on v7.

Please advise.

Thank you.

Premium customer - Support ID: 96081

**Comment by michelengelen** [comment]
> Hi @MBilalShafi
> 
> Server side row grouping was fine on MUI Data Grid v7 (Grouping on single and multiple columns).
> 
> Upon upgrading to v8 (8.5.2 -- also tried 8.0.0), the following was observed:
> 
> Screenshot 2025-06-16 230250
> 
> `Multiple` is the first level grouping.
> 
> When clicking on it to expand, the second level groups do appear (`Ahmed`, `Client Anastasia` and `Guest`).
> 
> However:
> 
> * Second level groups are not indented **(Issue 1)**
> * First level group clicked is not expanded (chevron is still pointing to the right) **(Issue 2)**
> 
> Clicking the first level group `Multiple` again correctly shows the expanded state (chevron down) and indent for second level groups.
> 
> When trying to click and expand other first level groups, the following error appears **(Issue 3)**:
> 
> ```
> Uncaught TypeError: Cannot read properties of undefined (reading 'depth')
>     at gridFilterSelector.js:92:39
>     at Array.reduce (<anonymous>)
>     at gridFilterSelector.js:90:30
>     at recomputationWrapper (createSelectorCreator.ts:392:63)
>     at memoized (lruMemoize.ts:217:20)
>     at dependenciesChecker (createSelectorCreator.ts:412:39)
>     at memoized (weakMapMemoize.js:100:21)
>     at Object.selector (createSelector.js:82:14)
>     at useGridSelector.js:59:37
>     at Store.js:19:35
> ```
> 
> These issues never occurred on v7.
> 
> Please advise.
> 
> Thank you.
> 
> Premium customer - Support ID: 96081

Hey @tommy-wl could you please open a new issue with this information?
And please include a minimal reproduction of the bug we can look at.
